src/coord/src/persistence.rs

```diff
@@ -0,0 +1,192 @@
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::PathBuf;
+
+use log::{debug, error};
+use timely::progress::Antichain;
+use tokio::sync::mpsc;
+
+use dataflow_types::Update;
+use expr::GlobalId;
+use repr::Timestamp;
+use storage::{Compacter, CompacterMessage, Message, Trace, WriteAheadLogs};
+
+#[derive(Clone, Debug)]
+pub struct PersistenceConfig {
+    pub wals_path: PathBuf,
+    pub traces_path: PathBuf,
+}
+
+pub struct PersistentTables {
+    wals: WriteAheadLogs,
+    compacter_tx: mpsc::UnboundedSender<CompacterMessage>,
+    wals_path: PathBuf,
+    traces_path: PathBuf,
+    disabled: bool,
+}
+
+impl PersistentTables {
+    pub fn new(config: &PersistenceConfig) -> Option<Self> {
+        let (compacter_tx, compacter_rx) = mpsc::unbounded_channel();
+        let mut compacter = match Compacter::new(
+            compacter_rx,
+            config.traces_path.clone(),
+            config.wals_path.clone(),
+        ) {
+            Ok(compacter) => compacter,
+            Err(e) => {
+                error!(
+                    "Encountered error while trying to initialize compacter task: {}",
+                    e
+                );
+                return None;
+            }
+        };
+        tokio::spawn(async move { compacter.run().await });
+        let wals = match WriteAheadLogs::new(config.wals_path.clone()) {
+            Ok(wals) => wals,
+            Err(e) => {
+                error!(
+                    "encountered error while trying to initialize write-ahead log: {}",
+                    e
+                );
+                return None;
+            }
+        };
+
+        Some(Self {
+            wals,
+            compacter_tx,
+            wals_path: config.wals_path.clone(),
+            traces_path: config.traces_path.clone(),
+            disabled: false,
+        })
+    }
+
+    pub fn resume(&mut self, id: GlobalId) -> Option<Vec<Message>> {
+        if self.disabled {
+            return None;
+        }
+
+        let persisted_trace = match Trace::resume(id, &self.traces_path, &self.wals_path) {
+            Ok(trace) => trace,
+            Err(e) => {
+                error!(
+                    "encountered error while trying to restart table {} {}",
+                    id, e
+                );
+                self.disabled = true;
+                return None;
+            }
+        };
+
+        let messages = match persisted_trace.read() {
+            Ok(messages) => Some(messages),
+            Err(e) => {
+                error!(
+                    "encountered error trying to read from persisted table {} {}",
+                    id, e
+                );
+                None
+            }
+        };
+
+        self.compacter_tx
+            .send(CompacterMessage::Resume(id, persisted_trace))
```

Comment:
It seems that re-reading the stored rows is governed by some sort of a flow control mechanism that makes it painfully slow for the data to arrive back into the tables. I think any database recovery operations should be exempt from any flow control constraints so that they can proceed at maximum speed. Right now this seems to be gated at 1000 rows/second or thereabout, which is sub-optimal.

Reply:
So I think I know what this might be -- did you wait for the log segments to get compacted into one big batch? I used the same script as above to generate 100k distinct inserts (thank you for that btw) and I waited until I had one large batch file
```
altaria-2:materialize Test$ ls -lrth mzdata/table_traces/u3/
total 6396008
-rw-r--r--  1 Test  staff   3.0G Mar  2 11:52 batch-0-1614703827970
altaria-2:materialize Test$
```

which took about 5 minutes but then when I restarted the restart took < 5 seconds. I never observed a count(*) < 100k on the relation

Reply:
I did not wait for the compaction to run prior to killing and restarting Materialize.